ahnlich/ai/src/engine/ai/models.rs

```diff
@@ -117,8 +241,146 @@ impl ModelInfo {
             input_type: model.input_type(),
             embedding_size: model.embedding_size(),
             max_input_tokens: model.max_input_token(),
-            max_image_dimensions: model.max_image_dimensions(),
+            expected_image_dimensions: model.expected_image_dimensions(),
             description: model.model_description(),
         }
     }
 }
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum InputAction {
+    Query,
+    Index,
+}
+
+impl fmt::Display for InputAction {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Query => write!(f, "query"),
+            Self::Index => write!(f, "index"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum ModelInput {
+    Text(String),
+    Image(ImageArray),
+}
+
+#[derive(Debug, Clone, PartialEq, Hash, Eq)]
+pub struct ImageArray {
+    array: Array<u8, Ix3>,
+    bytes: Vec<u8>,
+}
+
+impl ImageArray {
```

Comment:
Nice work with this internal abstraction.... makes a lot of sense to reason about it this way